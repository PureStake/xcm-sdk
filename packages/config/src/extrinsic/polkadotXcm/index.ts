export * from './polkadotXcm';
export * from './polkadotXcm.interfaces';
