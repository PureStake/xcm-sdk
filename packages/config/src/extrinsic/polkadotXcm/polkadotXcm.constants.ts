export enum PolkadotXcmExtrinsic {
  LimitedReserveTransferAssets = 'limitedReserveTransferAssets',
  LimitedReserveWithdrawAssets = 'limitedReserveWithdrawAssets',
}

export enum PolkadotXcmExtrinsicSuccessEvent {
  Attempted = 'Attempted',
}
