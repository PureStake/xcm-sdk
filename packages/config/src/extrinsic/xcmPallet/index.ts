export * from './xcmPallet';
export * from './xcmPallet.interfaces';
