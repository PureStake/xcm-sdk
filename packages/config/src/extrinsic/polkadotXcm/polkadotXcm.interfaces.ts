import { AssetId } from '../../types';
import { Parents } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';

export interface PolkadotXcmPallet {
  pallet: ExtrinsicPallet.PolkadotXcm;
  extrinsic: PolkadotXcmExtrinsic;
  successEvent: PolkadotXcmExtrinsicSuccessEvent;
  getParams: (account: string, amount: bigint) => PolkadotXcmPalletParams;
}

export type PolkadotXcmPalletParams = [
  /**
   * destination
   */
  {
    V1: {
      parents: Parents;
      interior: {
        X1: {
          Parachain: number;
        };
      };
    };
  },
  /**
   * beneficiary
   */
  {
    V1: {
      parents: 0;
      interior: {
        X1: {
          AccountKey20: {
            network: 'Any';
            /**
             * account
             */
            key: string;
          };
        };
      };
    };
  },
  /**
   * asset
   */
  PolkadotXcmAssetParam,
  /**
   * fee
   */
  0,
  /**
   * weight
   */
  {
    Limited: number;
  },
];

export type PolkadotXcmAssetParam =
  | PolkadotXcmAssetParamV0
  | PolkadotXcmAssetParamV1;

export interface PolkadotXcmAssetParamV0 {
  V0: [
    {
      ConcreteFungible: {
        id: 'Null';
        amount: bigint;
      };
    },
  ];
}

export interface PolkadotXcmAssetParamV1 {
  V1: [
    {
      id: {
        Concrete: {
          parents: Parents;
          interior:
            | 'Here'
            | PolkadotXcmAssetParamV1InteriorX1
            | PolkadotXcmAssetParamV1InteriorX2;
        };
      };
      fun: {
        Fungible: bigint;
      };
    },
  ];
}

export interface PolkadotXcmAssetParamV1InteriorX1 {
  X1: {
    PalletInstance: number;
  };
}

export interface PolkadotXcmAssetParamV1InteriorX2 {
  X2:
    | [
        {
          PalletInstance: number;
        },
        {
          GeneralIndex: AssetId;
        },
      ]
    | [
        {
          Parachain: number;
        },
        {
          PalletInstance: number;
        },
      ];
}
