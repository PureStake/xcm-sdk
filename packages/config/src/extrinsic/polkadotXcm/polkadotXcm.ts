import { AssetConfig, ChainConfig, MoonChainConfig } from '../../interfaces';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';
import { getCreateExtrinsic } from './polkadotXcm.util';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function polkadotXcm<Assets>(chain: MoonChainConfig) {
  return {
    limitedReserveTransferAssets: () =>
      limitedReserveTransferAssets<Assets>(chain),
    limitedReserveWithdrawAssets: () => limitedReserveWithdrawAssets(chain),
  };
}

function limitedReserveTransferAssets<Assets>(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveTransferAssets,
          event,
          chain,
          origin,
        );

        return {
          V0: () =>
            createExtrinsic((amount) => ({
              V0: [
                {
                  ConcreteFungible: {
                    id: 'Null',
                    amount,
                  },
                },
              ],
            })),
          V1: () => ({
            here: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: 'Here',
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X1: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: 5,
                          },
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X2: (palletInstance: number, asset: AssetConfig<Assets>) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              // TODO: throw an error if id is not defined or fix types
                              GeneralIndex: asset.originAssetId || 0,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}

function limitedReserveWithdrawAssets(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveWithdrawAssets,
          event,
          chain,
          origin,
        );

        return {
          V1: () => ({
            X2: (palletInstance: number) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 1,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              Parachain: chain.parachainId,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}
