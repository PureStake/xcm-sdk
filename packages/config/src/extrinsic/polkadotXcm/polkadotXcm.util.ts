import { ChainKey } from '../../constants';
import { Chain, MoonChain } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';
import {
  PolkadotXcmAssetParam,
  PolkadotXcmPallet,
} from './polkadotXcm.interfaces';

export function getCreateExtrinsic<ChainKeys extends ChainKey>(
  extrinsic: PolkadotXcmExtrinsic,
  event: PolkadotXcmExtrinsicSuccessEvent,
  config: MoonChain,
  origin: Chain<ChainKeys>,
) {
  return (
    getAsset: (amount: bigint) => PolkadotXcmAssetParam,
  ): PolkadotXcmPallet => ({
    pallet: ExtrinsicPallet.PolkadotXcm,
    extrinsic,
    successEvent: event,
    getParams: (account, amount) => [
      {
        V1: {
          parents: 1,
          interior: {
            X1: {
              Parachain: config.parachainId,
            },
          },
        },
      },
      {
        V1: {
          parents: 0,
          interior: {
            X1: {
              AccountKey20: {
                network: 'Any',
                key: account,
              },
            },
          },
        },
      },
      getAsset(amount),
      0,
      {
        Limited: origin.weight,
      },
    ],
  });
}
