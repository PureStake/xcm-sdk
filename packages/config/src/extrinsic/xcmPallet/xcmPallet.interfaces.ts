import {
  Pallet,
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from '../extrinsic.constants';
import { PolkadotXcmPalletParams } from '../extrinsic.interfaces';

export interface XcmPallet {
  pallet: Pallet.XcmPallet;
  extrinsic: PolkadotXcmExtrinsic.LimitedReserveTransferAssets;
  successEvent: PolkadotXcmExtrinsicSuccessEvent;
  params: PolkadotXcmPalletParams;
}
