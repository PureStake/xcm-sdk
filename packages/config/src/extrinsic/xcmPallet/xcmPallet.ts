import { ChainKey } from '../../constants';
import { Chain, MoonChain } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  getCreateExtrinsic,
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
  XcmMLVersion,
} from '../polkadotXcm';
import { XcmPallet } from './xcmPallet.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function xcmPallet<ChainKeys extends ChainKey>(chain: MoonChain) {
  return {
    limitedReserveTransferAssets: () =>
      limitedReserveTransferAssets<ChainKeys>(chain),
  };
}

function limitedReserveTransferAssets<ChainKeys extends ChainKey>(
  chain: MoonChain,
) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: Chain<ChainKeys>): XcmPallet => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveTransferAssets,
          event,
          chain,
          origin,
          0,
        );

        return {
          ...createExtrinsic(
            (amount) => ({
              V0: [
                {
                  ConcreteFungible: {
                    id: 'Null',
                    amount,
                  },
                },
              ],
            }),
            XcmMLVersion.v0,
          ),
          pallet: ExtrinsicPallet.XcmPallet,
        };
      },
    }),
  };
}
