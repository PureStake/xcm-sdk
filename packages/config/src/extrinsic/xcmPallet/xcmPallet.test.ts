import { Chain, MoonChain } from '../../constants';
import { ChainConfig, MoonChainConfig } from '../../interfaces';
import { PolkadotXcmExtrinsicSuccessEvent } from '../polkadotXcm';
import { xcmPallet } from './xcmPallet';

describe('xcmPallet', () => {
  const amount = 1000n;
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const chain: MoonChainConfig = {
    chain: MoonChain.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
    decimals: 10,
    unitsPerSecond: 100n,
  };
  const origin: ChainConfig = {
    chain: Chain.AlphanetRelay,
    name: 'Alphanet Relay',
    ws: 'wss://frag-moonbase-relay-rpc-ws.g.moonbase.moonbeam.network',
    weight: 1_000_000_000,
    parachainId: 0,
  };
  const extrinsic = xcmPallet(chain);

  describe('limitedReserveTransferAssets', () => {
    const cfg = extrinsic
      .limitedReserveTransferAssets()
      .successEvent(PolkadotXcmExtrinsicSuccessEvent.Attempted)
      .origin(origin);

    it('should be correct config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get correct params', () => {
      expect(cfg.getParams(account, amount)).toMatchSnapshot();
    });
  });
});
