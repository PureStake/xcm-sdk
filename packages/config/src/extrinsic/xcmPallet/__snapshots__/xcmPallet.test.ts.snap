// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xcmPallet limitedReserveTransferAssets v1v2 should be correct config 1`] = `
{
  "extrinsic": "limitedReserveTransferAssets",
  "getParams": [Function],
  "pallet": "xcmPallet",
  "successEvent": "Attempted",
}
`;

exports[`xcmPallet limitedReserveTransferAssets v1v2 should get correct params 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 1000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;
